### PR TITLE
feat: update tags on mailchimp add_contact

### PR DESCRIPTION
**Pull request by leogermani**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Adds the possibility to inform tags to be added and tags to be removed in the `add_contact` method for mailchimp

### How to test the changes in this Pull Request:

1. Open a terminal and test add a contact using the new methods
2. Example 

```
$mc = Newspack_Newsletters_Mailchimp::instance();
$mc->add_contact( [ 'email' => 'tesint@tesintgcom', 'name' => 'Test contact', 'add_tags' => ['tag-1','tag-2'] ], '__LIST_ID__' );
```
Close shell session and open a new one:
```
$mc = Newspack_Newsletters_Mailchimp::instance();
$mc->add_contact( [ 'email' => 'tesint@tesintgcom', 'name' => 'Test contact', 'remove_tags' => ['tag-2'] ], '__LIST_ID__' );
```
4. Check in the Mailchimp dashboard to see if the updates were done correctly

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php [hunk 1]
```diff
@@ -1176,6 +1185,23 @@ function( $a, $b ) {
...
+
+			// Add tags if any were informed.
+			if ( ! empty( $contact['add_tags'] ) ) {
+				foreach ( $contact['add_tags'] as $tag ) {
+					$tag_id = $this->get_tag_id( $tag, true, $list_id );
+					$this->add_tag_to_contact( $email_address, $tag_id, $list_id );
+				}
+			}
+			// Remove tags if any were informed.
+			if ( ! empty( $contact['remove_tags'] ) ) {
```
Comment (chickenn00dle) [inline includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php:1197]: When I run the following commands:

```
$mc = Newspack_Newsletters_Mailchimp::instance();
$mc->add_contact( [ 'email' => 'raz+shell@automattic.com', 'name' => 'Test contact', 'add_tags' => ['tag-1','tag-2'] ], 'LISTID' );
$mc->add_contact( [ 'email' => 'raz+shell@automattic.com', 'name' => 'Test contact', 'remove_tags' => ['tag-2'] ], 'LISTID' );
```

The contact is created, however the `tag-2` is not removed in MC:
Screenshot 2024-02-13 at 17 05 51

File: includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php [hunk 2]
```diff
@@ -1342,17 +1368,34 @@ public function get_contact_data( $email, $return_details = false ) {
...
+	/**
+	 * A memoized version of the get_contact_data method.
+	 *
+	 * To be used in methods where it's not critical that we have the most updated version of a contact
+	 *
+	 * @param string $email The contact email.
+	 * @return array|WP_Error Response or error if contact was not found.
+	 */
+	public function get_cached_contact_data( $email ) {
+		if ( isset( $this->contacts_data[ $email ] ) && ! is_wp_error( $this->contacts_data[ $email ] ) ) {
```
Comment (chickenn00dle) [inline includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php:1384]: Feels a bit strange to allow errors to exist in contact data cache. Should we be handling these errors before they are populated here, and maybe filtering these entries out?

**Review by chickenn00dle — CHANGES_REQUESTED** [review]

**Review by chickenn00dle — COMMENTED** [review]
Ah. One more thing I forgot to include in the review. I am seeing the following warning for this line: https://github.com/Automattic/newspack-newsletters/blob/cae7510ad2e5e29c23a9a7abaab2299881e1454c/includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php#L1364

```
PHP Notice:  Undefined index: interests in /Users/raz/Sites/newspack/newsletters/includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php on line 1364
```

Seems unrelated to these changes, but maybe we can use this PR to sneak in a fix for it as well?

File: includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php (hunk 2 above)
Comment (leogermani) [reply includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php:1384]: > Should we be handling these errors before they are populated here, and maybe filtering these entries out?

We actually are, if you look at the method above. I added this check before realizing it. I'll remove this check :+1:

Comment (leogermani) [reply includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php]: removed in https://github.com/Automattic/newspack-newsletters/pull/1433/commits/61819ec55dc9c1e22c1c25d77e8cf5558f043f13

**Comment by leogermani** [comment]
> Ah. One more thing I forgot to include in the review. I am seeing the following warning for this line:
> 
> https://github.com/Automattic/newspack-newsletters/blob/cae7510ad2e5e29c23a9a7abaab2299881e1454c/includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php#L1364
> 
> ```
> PHP Notice:  Undefined index: interests in /Users/raz/Sites/newspack/newsletters/includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php on line 1364
> ```
> 
> Seems unrelated to these changes, but maybe we can use this PR to sneak in a fix for it as well?

I can't see those but I guess it can happen if there are no groups in your Audience?... 

Addressed in https://github.com/Automattic/newspack-newsletters/pull/1433/commits/4a56f3e710175bc561662d0a7e94ea7f353cfe24

File: includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php (hunk 1 above)
Comment (chickenn00dle) [reply includes/service-providers/mailchimp/class-newspack-newsletters-mailchimp.php:1197]: Aha! That did the trick 👍

**Review by chickenn00dle — APPROVED** [review]
Great! Thanks for the quick turnaround. This looks good now 👍

**Comment by leogermani** [comment]
I want to wai just a little to merge this as I'm working on the methods that will use this new feature. There might be some little adjustments to make

**Comment by dkoo** [comment]
Superseded by #1471

**Pull request closed without merging**